### PR TITLE
Add Instance Segmentation Transforms

### PR DESCRIPTION
**Pull request by yutong-xiang-97**
## What has changed and why?

This PR stacks on top of #835.

It adds LT-DETR instance segmentation transform support on top of the object detection transform refactor:

- add instance segmentation dataset mask handling
- add LT-DETR instance segmentation transform wiring
- add a reusable `ltdetr_transforms` instance segmentation transform module
- add focused transform coverage for the new instance segmentation path

## How has it been tested?

Unit tests.

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [ ] Yes
- [x] Not needed (internal change)

## Did you update the documentation?

- [ ] Yes
- [x] Not needed (internal change without effects for user)

### CODE REVIEW

File: src/lightly_train/_transforms/ltdetr_transforms/instance_segmentation.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        image = input["image"]
+        bboxes = input["bboxes"]
+        class_labels = input["class_labels"]
+        binary_masks = input["binary_masks"]
+        indices = np.arange(len(bboxes), dtype=np.int64)
+
+        if self._should_apply_mosaic():
+            image, bboxes, class_labels, binary_masks_opt = self.mosaic(  # type: ignore[misc]
+                image, bboxes, class_labels, binary_masks
+            )
```
[the remark this answers is not in the document]
Comment (yutong-xiang-97) [reply src/lightly_train/_transforms/ltdetr_transforms/instance_segmentation.py]: I raised an error instead.

File: src/lightly_train/_transforms/ltdetr_transforms/utils.py [hunk 2]
```diff
@@ -134,13 +129,13 @@ def resolve_ltdetr_step_schedule_for_augmentation(
                 f"step_stop ({step_schedule.step_stop}) <= 0. "
                 "Disabling scale_jitter for this run."
             )
-            args.scale_jitter = None
```
Comment (gabrielfruet) [inline src/lightly_train/_transforms/ltdetr_transforms/utils.py:137]: Any reason why using seattr here instead of just setting the attribute?

File: src/lightly_train/_commands/train_task.py [hunk 3]
```diff
@@ -1497,14 +1497,14 @@ def _train_task_from_config(config: TrainTaskConfig) -> None:
             model_name=config.model,
             val_transform_args=val_transform_args,
         )
-        train_collate_fn = train_dataset.batch_collate_fn_cls(
+        train_dataset.set_transform(train_transform)
+        val_dataset.set_transform(val_transform)
+        train_collate_fn = train_dataset.get_batch_collate_fn_cls()(
```
Comment (gabrielfruet) [inline src/lightly_train/_commands/train_task.py:1502]: Why we have this method instead of direct access to the batch_collate_fn_cls?

File: src/lightly_train/_transforms/ltdetr_transforms/instance_segmentation.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+    scale_jitter: ScaleJitterArgs | None = None
+    resize: ResizeArgs | None
+    bbox_params: BboxParams | None
+    normalize: NormalizeArgs | Literal["auto"] | None
+
+    # Necessary for BboxParams, which are not serializable by pydantic.
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+
+class _MaskAwareStepScheduledCompose(StepScheduledCompose):
```
Comment (gabrielfruet) [inline src/lightly_train/_transforms/ltdetr_transforms/instance_segmentation.py:101]: Why we don't just make `StepScheduledCompose` MaskAware?

**Review by gabrielfruet — COMMENTED** [review]
In general LGTM, I just have some questions

File: src/lightly_train/_commands/train_task.py (hunk 3 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_commands/train_task.py:1502]: Good point. It is because instance segmentation now should select the collate classes based on the model architecture, but I noticed that this particular implementations assumes EoMT as the default. I will make `get_batch_collate_fn_cls()` return EoMT or LT-DETR purely based on `self.transform` but not remove it.

File: src/lightly_train/_transforms/ltdetr_transforms/utils.py (hunk 2 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_transforms/ltdetr_transforms/utils.py:137]: In utils.py, `args` is typed as `LTDETRTransformArgs`, and that is a `Protocol `exposing `scale_jitter` as a read-only `@property`. This will make mypy reject `
args.scale_jitter = None` because the protocol does not define a setter. Instead, `setattr(args, "scale_jitter", None)` sidesteps that static typing limitation while still doing the runtime mutation on the concrete Pydantic model.

File: src/lightly_train/_transforms/ltdetr_transforms/instance_segmentation.py (hunk 4 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_transforms/ltdetr_transforms/instance_segmentation.py:101]: `StepScheduledCompose` is currently the shared LT-DETR scheduling/cache component. Object detection uses it with image + bbox + class labels only, while instance segmentation must pass the masks through the compose and `ToTensorV2` must convert image and masks together.
Making the parent mask-aware would either add mask-specific behavior to the object-detection path, or require flags/conditionals inside the shared component. Neither makes sense for me.

Comment (gabrielfruet) [reply src/lightly_train/_transforms/ltdetr_transforms/instance_segmentation.py:101]: Based on the class implementation, we would only need to add ToTensorV2() to the pipeline, which seems very reasonable to me. ToTensorV2() shouldn't affect the object detection path I think

Comment (yutong-xiang-97) [reply src/lightly_train/_transforms/ltdetr_transforms/instance_segmentation.py:101]: Nope adding `ToTensorV2` to object detection would actually change output contract of the transform.

Object detection’s `StepScheduledCompose` returns numpy-style Albumentations outputs, and the object-detection collate keeps working with it while it applies batch-level augmentations. Tensor conversion happens only at the end of the collate path. If the parent became mask-aware by always adding `ToTensorV2`, object detection would start receiving torch tensors from the sample transform. That has a few consequences. Most notably, you will receive CHW tensors instead of HWC numpy arrays; also, the shape and dtype assumptions in mixup/copyblend/scale-jitter could change;

For instance segmentation, on the other hand, appending `ToTensorV2` at sample level is specifically needed because Albumentations converts image and masks together in one compose call, and the instance transform then indexes/stacks the returned masks as tensors.

**Review by gabrielfruet — APPROVED** [review]

File: src/lightly_train/_transforms/ltdetr_transforms/instance_segmentation.py (hunk 4 above)
Comment (gabrielfruet) [reply src/lightly_train/_transforms/ltdetr_transforms/instance_segmentation.py:101]: Makes sense!
